### DISCUSSION TITLE
Dear imgui as an webapplication

### DISCUSSION THREAD

**Discussion by scinteeb** [ocornut/imgui#5428] [New Users Build/Link/Run issues ONLY!]
Hello,
  I am planning to write an webapplication for an embedded device that will allow to login on the device, make settings and visualize several parameters in real time using gauges and charts. Dear imgui (even it is promoted more for gaming), when compiled with emscripten, seems to be a very good fit for what I plan to do. However, this imgui and wasm is new for me and so far I couldn't find what is the best way to have a reliable and fast communication between the imgui client and some backend service that will facilitate this data exchange. gRPC, protobuf, WebSockets, DDS? Is there anything already available/integrated? I just don't want to reinvent the wheel.
Probably the question was more suitable for a thread related to WebAssembly. :(

Anyway, thanks and good luck,
Bob.

**Comment by PathogenDavid** [comment]
I don't think there's really any one method of communication that is best here (definitely nothing that's already integrated), it mostly comes down to your own preference and what your team is comfortable working with.

Depending on your exact needs, you might actually check out [one of the various third-party remoting frameworks](https://github.com/ocornut/imgui/wiki/Useful-Extensions#remoting). This skips needing to do your own data exchange entirely because UI submission happens on the embedded device and all the client does is the final rendering of the vertex data. (Downside of this approach is you'll end up sending a lot more data to the client, upside is you don't have to deal with defining a protocol between the client and server.)

**Reply by scinteeb** [reply]
Thank you for your feedback. I believe that the remoting is not a good approach. As I said, my intention is to create the webapplication for an embedded device. So if I am doing a plotting from a number of clients, I can 'kill' the device just with rendering UIs. A pub/sub mechanism is more appropriate as it will feed to the clients only the data needed to be monitored.

**Reply by PathogenDavid** [reply]
Agreed, if you want to support multiple clients with their own separate UIs the remoting strategy is likely not a viable choice.

**Comment by abinash18** [comment]
Hey, this is a good question and unfortunately, I found a solution too late to help you since I just started my work term and tried to do this with an embedded device. the best solution I found is WebSockets!!!! full-duplex asynchronous with emscripten compiled js and very light weight on the embedded device.

**Reply by abinash18** [reply]
To anyone who comes across this post and is confused as to how to do this, you can email me if you want. But here is the general process I went through took me 1 week to figure out and I think this will help everyone in need:
1: there is 1 good library I found for handling web sockets on Arduino. Arduino not bare-metal, WebSockets2_Generic just seems poorly desined imo idk who that guy is but he doesn't seem to have any good documentation for my need and his library is too big and depends on other things he made and whatever. it's just too big and spaghetti imo so I used a library called mWebSockets in Arduino, its simple and small and does the one thing I need it to.

2: Arduino Json to send message bodies and commands both ways. 

3: emscripten with imgui. Right now I am using hello_imgui which is another pre-designed framework to compile to literally every relevant platform I can think of. 

4: make use of the ability of emscripten to make WebWorkers which run asynchronously and from what I have read can also transfer data asynchronously to the imgui process on the browser.

Remarks:
- I did not use multithreading as I didnt see a good reason for it on the web version. since in the web people usually just use 2 different workers.
- WebSockets offer SSL connection. although imo u should never connect an embedded device to the internet especially if you are making a product that is for industrial use without a specific requirement. It should always be behind a proxy or a VPN.

**Reply by abinash18** [reply]
I also used the Wiz550io shield on a blackpill f4c11e board
so mWebSockets lets you set up ur own ethernet or wifi and dosnt force you like WebSockets Generic and also dosnt download like a million F'in dependecies.

**Reply by abinash18** [reply]
https://github.com/pthom/hello_imgui
This deserves so many stars

**Reply by abinash18** [reply]
I lied use threads its easier lmao.
